src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,258 @@
+        }
+        if self.unique_range.end >= new_idx {
+            self.unique_range.end += 1;
+        }
+        if new.perm == Permission::Unique {
+            // Make sure the possibly-unique range contains the new borrow
+            self.unique_range.start = self.unique_range.start.min(new_idx);
+            self.unique_range.end = self.unique_range.end.max(new_idx + 1);
+        }
+
+        // The above insert changes the meaning of every index in the cache >= new_idx, so now
+        // we need to find every one of those indexes and increment it.
+        for idx in &mut self.cache.idx {
+            if *idx >= new_idx {
+                *idx += 1;
+            }
+        }
+
+        if self.len() < CACHE_LEN {
+            self.cache.tags[self.len()] = new.tag;
+            self.cache.idx[self.len()] = new_idx;
+        }
+
+        #[cfg(feature = "expensive-debug-assertions")]
+        self.verify_cache_consistency();
+    }
+
+    pub fn new(item: Item) -> Self {
+        Stack {
+            borrows: vec![item],
+            cache: StackCache { idx: [0; CACHE_LEN], tags: [item.tag; CACHE_LEN] },
+            unique_range: if item.perm == Permission::Unique { 0..1 } else { 0..0 },
+        }
+    }
+
+    pub fn get(&self, idx: usize) -> Option<Item> {
+        self.borrows.get(idx).cloned()
+    }
+
+    pub fn len(&self) -> usize {
+        self.borrows.len()
+    }
+
+    pub fn disable_uniques_above<V: FnMut(Item) -> crate::InterpResult<'static>>(
+        &mut self,
+        granting_idx: usize,
+        mut visitor: V,
+    ) -> crate::InterpResult<'static> {
+        if granting_idx < self.unique_range.end {
+            // add 1 so we don't disable the granting item
+            let lower = self.unique_range.start.max(granting_idx + 1);
+            let upper = (self.unique_range.end + 1).min(self.borrows.len());
+            for item in &mut self.borrows[lower..upper] {
+                if item.perm == Permission::Unique {
+                    log::trace!("access: disabling item {:?}", item);
+                    visitor(item.clone())?;
+                    item.perm = Permission::Disabled;
+                }
+            }
+        }
+
+        if granting_idx + 1 < self.unique_range.start {
+            // We disabled all Unique items
+            self.unique_range.start = 0;
+            self.unique_range.end = 0;
+        } else {
+            // Truncate the range to granting_idx. This is + 2 because we are only removing
+            // elements after granting_idx, and this range does not include the end.
+            self.unique_range.end = self.unique_range.end.min(granting_idx + 2);
+        }
+
+        #[cfg(feature = "expensive-debug-assertions")]
+        self.verify_cache_consistency();
+
+        Ok(())
+    }
+
+    pub fn drain_rev(&mut self, range: std::ops::Range<usize>) -> DrainRev<'_> {
+        DrainRev { start: range.start, stack: self, cursor: range.end - 1 }
+    }
```

Comment:
This is only ever used with a range to `self.len()`, maybe just pass a start and compute the end here? the `- 1` offset made me wonder if it could bite us somewhere.